pkg/apis/experimental/types.go

```diff
@@ -461,3 +461,75 @@ type JobCondition struct {
 	// Human readable message indicating details about last transition.
 	Message string `json:"message,omitempty"`
 }
+
+// IngressPoint encapsulates the inputs needed to connect an alias to endpoints.
+// It represents a rule collection of inbound connections from the external network
+// that would be satisfied by a load balancer.
+type IngressPoint struct {
+	api.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the IngressPoint.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Spec IngressPointSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the IngressPoint.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Status IngressPointStatus `json:"status,omitempty"`
+}
+
+// IngressPointList is a collection of IngressPoints,
+type IngressPointList struct {
+	api.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	api.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of ingressPoints.
+	Items []IngressPoint `json:"items"`
+}
+
+// ServiceRef represents a reference to a single service:port.
+type ServiceRef struct {
+	// Define the name of the referenced service.
+	Name string `json:"name,omitempty"`
+
+	// Define the namespace of the referenced service.
+	Namespace string `json:"namespace,omitempty"`
+
+	// Define the service's port which the http traffic will be forwarded to.
+	Port int `json:"port,omitempty"`
+}
+
+// PathRef represents a single load balancer rule of forwarding the specified url to the related backend service.
+type PathRef struct {
+	// Domain describe http url's domain.
+	// Examples: "foo.example.com", "bar.example.com".
+	Domain string `json:"domain,omitempty"`
+
+	// Path defines http url's path.
+	// Examples: "/images", "/images/*".
+	Path string `json:"path,omitempty"`
+
+	// The ServiceRef indicates the unified service by service name, namespace and port.
+	Service ServiceRef `json:"service,omitempty"`
+}
+
+// IngressPointSpec describes the ingressPoint the user wishes to exist.
+type IngressPointSpec struct {
+	// Alias/Dns that points to the services, can be host or host:port.
+	// Examples: "www.example.com", "www.example.com:8080".
```

Comment:
In your example I think you reversed IngressSpec and IngressStatus

IngressRules or IngressRoutes is the best I can come up with sa an
alternative to IngressMap.  We *could* just call it "mappings
[]IngressMapping".

If the object is IngressSubdomain, I think the field would be "Name" not
"SubDomain" (it's the name of the object).  "subdomains[0].subdomain"
doesn't have parallels in our other objects (pod.spec.containers[0].name,
not pod.spec.containers[0].container)

On Wed, Sep 16, 2015 at 2:24 PM, Prashanth B <notifications@github.com>
wrote:

> In pkg/apis/experimental/types.go
> <https://github.com/kubernetes/kubernetes/pull/13947#discussion_r39666680>
> :
>
> > +	// Domain describe http url's domain.
> > +	// Examples: "foo.example.com", "bar.example.com".
> > +	Domain string `json:"domain,omitempty"`
> > +
> > +	// Path defines http url's path.
> > +	// Examples: "/images", "/images/*".
> > +	Path string `json:"path,omitempty"`
> > +
> > +	// The ServiceRef indicates the unified service by service name, namespace and port.
> > +	Service ServiceRef `json:"service,omitempty"`
> > +}
> > +
> > +// IngressPointSpec describes the ingressPoint the user wishes to exist.
> > +type IngressPointSpec struct {
> > +	// Alias/Dns that points to the services, can be host or host:port.
> > +	// Examples: "www.example.com", "www.example.com:8080".
>
> IngressBackend innocently equivocates service vs endpoint, but I don't
> think it's confusing.
>
> IngressSpec -> IngressMap -> IngressPathRule -> IngressBackend
>
> I think the IngressMap is too fuzzy, don't know what it contains (it has
> subdomains, not just paths->backends). We'll need to add comments to
> explain something we could otherwise explain via renaming. So what about:
>
> Ingress {
>   Spec IngressSpec
>   Status IngressStatus
> }
>
> IngressSpec {
>   LoadBalancers api.LoadBalancerStatus
> }
>
> IngressStatus {
>     Domain string,
>     SubDomains []IngressSubdomain
> }
>
> IngressSubdomain {
>     SubDomain string
>     PathRules []IngressPathRule
> }
>
> IngressPathRule {
>     PathExpr string
>     Backend IngressBackend
> }
>
> IngressBackend struct {
>     ServiceRef api.LocalObjectReference
>     ServicePort util.IntOrString
>     Protocol api.Protocol
> }
>
> Re splitting host and domain, I meant don't use host at all as a name. we
> can have multiple subdomains but only a single domain in one ingress.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13947/files#r39666680>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift